### PR TITLE
VC integration

### LINKED ISSUE #92: VC support
It would be great if `neotree` could (optionally) tie in to `vc`. See [this Emacs.SE question](http://emacs.stackexchange.com/q/9503/2264).

### PR DESCRIPTION
**Pull request by vermiculus**
screen shot 2015-03-22 at 9 11 31 pm

I think I just need some help finishing it up – I've done all I can think of. For some reason, certain files aren't being statused properly by `vc-state`. I can't think of what the problem is. I'm using the Emacs git repository as a test. (I don't have any SVN repos to test agains.)

You'll note that `.new-file` – instead of being green (`neo-vc-added-face`), it's `neo-vc-default-face`. `(vc-state "/Users/sean/.../.newfile")` is returning `nil` and I don't know why.

This will eventually resolve #92.

### CODE REVIEW

**Comment by jaypei** [comment]
Thanks!
Could you maybe merge into one commit?
It will make the PR cleanly.

**Comment by vermiculus** [comment]
Like that? First time doing a squash :)

**Comment by jaypei** [comment]
Excellent work!

**Comment by vermiculus** [comment]
Awesome stuff. Note that another customization option has been added, `neo-vc-state-char-alist`. Ideally, I'd like to shove everything into just two customization options – one to control which functionality is triggered and the other to control how the functionality behaves. To that end, do you know if there's any way to define 'anonymous faces' as part of a customization variable? (My gut tells me that it's not possible as of yet.)

File: neotree.el [hunk 1]
```diff
@@ -1046,18 +1167,48 @@ PATH is value."
...
+                           neo-file-link-face)
                    'neo-full-path node
                    'keymap nil)
     (neo-buffer--node-list-set nil node)
     (neo-buffer--newline-and-begin)))
 
+(defun neo-vc-for-node (node)
+  (let ((backend (vc-backend node)))
+    (let* ((vc-state (if backend
+                         (vc-state-refresh node backend)
```
Comment (vermiculus) [inline neotree.el]: By the way, this was the fix for the weird VC behavior I mentioned above. Turns out `vc-state` caches the return value.

**Comment by vermiculus** [comment]
Do you think it would be a good idea to add this hook?
```lisp
(add-hook 'write-file-functions (lambda () (when neo-vc-integration (neotree-refresh))))
```
I'm just not sure how kosher it is to alter (what I see as) user-level variables.

**Comment by dgutov** [comment]
I'd like to point out that doing one `(vc-state ...)` call for each tree node might be relatively slow.

It might be better to do what `diff-hl-dired` does and utilize the `dir-status-files` command: https://github.com/dgutov/diff-hl/blob/eb92f0e99c0de4d51b80d307fdb92e0533ff9e74/diff-hl-dired.el#L129-L134

On the flip side, its calling convention changed during Emacs 25 development, and it also had certain bugs fixed, one of which [needs working around](https://github.com/dgutov/diff-hl/blob/eb92f0e99c0de4d51b80d307fdb92e0533ff9e74/diff-hl-dired.el#L136-L144).

**Comment by vermiculus** [comment]
For some VCS, I suppose you're absolutely right. SVN for example might even need to run a network query for its statusing. I'll see what I can do about using this special function -- not too sure how it would work at the moment.

It'll have to wait until next week though -- I'm going to be missing-in-action for this weekend :)

**Comment by jaypei** [comment]
Hi vermiculus,

I am busy recently, but I will help you to merge this feature ASAP, because this is awesome.

Answer your first question, beacuse you defined the `neo-vc-added-face` was green,
but I guess the real status of this file is unregistered,
it has not use `git add` into cache stage.

Jaypei

**Comment by jaypei** [comment]
I tried svn repo, vc-stat no network requests occured.
I think dgutov's means is that the number of vc api calls should as few as possible.

If enable 'char and neo-theme is classic, the gap is too big between ICON and file-name.
I think `diff-hl-dired` is a very good reference, it used fringe.
diff-hl-dired

What do you think?


Jaypei

**Comment by vermiculus** [comment]
I'll have to learn how to place something in the fringe, but it won't be any trouble adding that as an indicator style. I want to keep the char style (because I'd use it personally), but I can definitely get behind the idea that it's not for everyone. I'll see what I can figure out this weekend :)

**Comment by jaypei** [comment]
I think the write-file-functions hook need more test, we can implement it and set disable by default.

**Comment by jaypei** [comment]
Do you think that put the status characters to the back of the file-name is a good idea?

**Comment by vermiculus** [comment]
I don't see why it wouldn't be, but I'm interested to know what you're thinking. I added the possibility in anticipation for using emacs over an ssh tunnel. Faces never worked for me through them (and TRAMP never worked well in those particular situations, either). Those days are gone for me, but I imagine there are others that are stuck in the same spot I was in.

**Comment by vermiculus** [comment]
Is there any reason this can't be merged and then improved at a later date?

**Comment by jaypei** [comment]
Thanks for your contribution. It is really awesome!

**Comment by vermiculus** [comment]
It was fun to do!

**Comment by Jumblemuddle** [comment]
This works great, except for visualizing ignored files with Git. Has anyone else run into issue with `vc-status` not working well with ignored files?
I've implemented quick workaround using `magit-ignored-files`, but it's far from ideal.

**Pull request merged**